Sources/SwiftSyntax/Syntax.swift

```diff
@@ -380,6 +384,57 @@ extension Syntax {
   }
 }
 
+/// Temporary non-owning Syntax.
+///
+/// This can be used for handling Syntax node without ARC traffic.
+struct UnownedSyntax {
+  var raw: RawSyntax
+  var info: Unmanaged<Syntax.Info>
+
+  @_transparent
+  init(_ node: __shared Syntax) {
+    self.raw = node.raw
+    self.info = .passUnretained(node.info.unsafelyUnwrapped)
+  }
+
+  /// Extract the Syntax value.
+  @inline(__always)
+  var value: Syntax {
+    Syntax(raw, info: info.takeUnretainedValue())
+  }
+
+  /// Get the parent of the Syntax value, but without retaining it.
+  @inline(__always)
+  var parent: UnownedSyntax? {
+    return info._withUnsafeGuaranteedRef {
+      switch $0.info.unsafelyUnwrapped {
+      case .nonRoot(let info):
+        return UnownedSyntax(info.parent)
+      case .root(_):
+        return nil
+      }
+    }
+  }
+
+  /// Temporarily use the Syntax value.
+  @inline(__always)
+  func withValue<T>(_ body: (Syntax) -> T) -> T {
+    info._withUnsafeGuaranteedRef {
+      body(Syntax(self.raw, info: $0))
+    }
+  }
+}
+
+/// Execute the `body` with ``UnownedSyntax`` of `node`.
+///
+/// This guarantees the life time of the `node` during the `body` is executed.
+@inline(__always)
+func withUnownedSyntax<T>(_ node: some SyntaxProtocol, _ body: (UnownedSyntax) -> T) -> T {
```

Comment:
Would this be better as a member on `SyntaxProtocol` so that you would do `self.withUnownedSyntax` instead of `withUnownedSyntax(self)`?